cjdns-core/src/encoding.rs

```diff
@@ -31,393 +32,614 @@
+        prefix_len: u8,
+        prefix: u32,
+    }
 
-        if encform.bit_count < 1 || encform.bit_count > 31 {
-            return Err(EncodingSchemeError::BadEncodingForm);
+    impl EncodingSchemeForm {
+        /// Instantiates `EncodingSchemeForm`.
+        ///
+        /// Returns an error in several situations:
+        /// * encoding `bit_count` value is out of valid range, which is 1..32
+        /// * `prefix_len` is too small for the provided `prefix`
+        pub fn try_new(bit_count: u8, prefix_len: u8, prefix: u32) -> Result<Self, FormValidationError> {
+            if bit_count == 0 || bit_count > 31 {
+                return Err(FormValidationError::BadBitCount);
+            }
+            if 2u32.pow(prefix_len as u32) <= 32 - prefix.leading_zeros() {
+                return Err(FormValidationError::InvalidPrefixData);
+            }
+            Ok(EncodingSchemeForm { bit_count, prefix_len, prefix })
         }
 
-        if encform.prefix_len > 0 {
-            acc = acc | encform.prefix as u64;
+        /// Returns encoding scheme form params in respected order:
+        /// * bit count;
+        /// * prefix length;
+        /// * prefix itself.
+        pub fn params(&self) -> (u8, u8, u32) {
+            (self.bit_count, self.prefix_len, self.prefix)
         }
 
-        acc = acc << 5;
-        acc = acc | encform.bit_count as u64;
+        /// As a scheme is represented as an array of **forms**, this function will tell you how many bits of
+        /// label space is occupied by a representation of a given form.
+        pub fn size_bits(&self) -> u8 {
+            self.bit_count + self.prefix_len
+        }
+    }
 
-        acc = acc << 5;
-        acc = acc | encform.prefix_len as u64;
+    impl EncodingScheme {
+        /// Instantiates `EncodingScheme`.
+        ///
+        /// Returns an error if forms validation failed. See `validate` function docs for more info.
+        pub fn try_new(forms: &[EncodingSchemeForm]) -> Result<Self, SchemeValidationError> {
+            let _ = Self::validate(forms)?;
+            Ok(Self(forms.to_vec()))
+        }
 
-        let bits_needed = 5 + 5 + encform.prefix_len;
-        // println!("[DEBUG] accum: {:064b}", accum64);
+        /// Validates encoding scheme.
+        ///
+        /// Returns an error in several situations:
+        /// * provided forms slice length is 0 or greater than 31
+        /// * `bit_count` value of any form is out of valid range - 1..32
+        /// * `prefix_len` value of any form is out if valid range - 1..32 (for multiple forms scheme)
+        /// * forms are not in ascending order by `bits_count` key
+        /// * bits size of a form is greater than 59 (for multiple forms scheme)
+        /// * forms with equal prefixes are in scheme
+        ///
+        /// Each returned value fully reflects error type.
+        pub fn validate(forms: &[EncodingSchemeForm]) -> Result<(), SchemeValidationError> {
```

Comment:
Why still public? Is it still necessary?

Reply:
the same as `size_bits`, which was public fn (now pub method) - we are unsure whether it would be used externally or not, but JS implementation exported this function.